File: product_of_all_other_numbers/product_of_all_other_numbers.py
def product_of_all_other_numbers(arr):
    products = [0] * len(arr)

    product_before = 1

    for i in range(len(arr)):
        products[i] = product_before
        product_before *= arr[i]

    product_after = 1

    for j in range(len(arr)-1, -1, -1):
        products[j] *= product_after
        product_after *= arr[j]

    return products

File: product_of_all_other_numbers/productOfAllOtherNumbers.ts
export const productOfAllOtherNumbers = (arr) => {
  const products = Array(arr.length).fill(0);

  let productBefore = 1;

  for (let i = 0; i < arr.length; i++) {
    products[i] = productBefore;
    productBefore *= arr[i];
  }

  let productAfter = 1;

  for (let j = arr.length - 1; j > -1; j--) {
    products[j] *= productAfter;
    productAfter *= arr[j];
  }

  return products;
};
